src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -547,7 +547,14 @@ CallStubHeader *CallStubGenerator::GenerateCallStub(MethodDesc *pMD, AllocMemTra
 
     _ASSERTE(pMD != NULL);
 
+    // Classes like System.String have special constructors that are fcalls. When invoking these, we need to override
+    //  HasThis (there's no thisref) and the return type (the return value is the new instance, not void).
+    bool isSpecialConstructor = pMD->IsCtor() && pMD->GetMethodTable()->HasComponentSize();
```

Comment:
The runtime uses a bit different condition to decide whether to clear the HasThis, it would be good to use the same one:
https://github.com/dotnet/runtime/blob/87b35f477c09f146107a6c11627317f6b15e9d3e/src/coreclr/vm/frames.cpp#L1518